langwatch/src/utils/executeEvalWorkflow.ts

```diff
@@ -0,0 +1,181 @@
+import { type NextApiRequest, type NextApiResponse } from "next";
+import { prisma } from "../server/db";
+import { getProjectModelProviders } from "../server/api/routers/modelProviders";
+import { type Workflow } from "../optimization_studio/types/dsl";
+import { nanoid } from "nanoid";
+import type { StudioClientEvent } from "../optimization_studio/types/events";
+import { type Edge, type Node } from "@xyflow/react";
+import { type MaybeStoredModelProvider } from "../server/modelProviders/registry";
+import { addEnvs } from "../optimization_studio/server/addEnvs";
+import { checkIsEvaluator } from "../optimization_studio/utils/nodeUtils";
+
+const getWorkFlow = (state: Workflow) => {
+  return {
+    workflow_id: state.workflow_id,
+    spec_version: state.spec_version,
+    name: state.name,
+    icon: state.icon,
+    description: state.description,
+    version: state.version,
+    default_llm: state.default_llm,
+    nodes: state.nodes,
+    edges: state.edges,
+    state: state.state,
+  };
+};
+
+const checkForRequiredInputs = (
+  publishedWorkflowVersion: Workflow,
+  body: Record<string, unknown>
+) => {
+  const bodyInputs = Object.keys(body);
+
+  const entryEdges = publishedWorkflowVersion?.edges.filter(
+    (edge: Edge) => edge.source === "entry"
+  );
+  const evaluators = publishedWorkflowVersion?.nodes.filter(checkIsEvaluator);
+
+  const entryInputs = entryEdges.filter(
+    (edge: Edge) =>
+      !evaluators?.some((evaluator: Node) => evaluator.id === edge.target)
+  );
```

Comment:
we do all this logic on the frontend when using a custom evaluator on the optimization studio as well right? Can you extract it to an optimization_studio/utils function used on both places?